src/tir/ir/data_type_rewriter.cc

```diff
@@ -107,6 +107,26 @@ Stmt DataTypeLegalizer::VisitStmt_(const AttrStmtNode* op) {
   return StmtExprMutator::VisitStmt_(op);
 }
 
+Stmt DataTypeLegalizer::VisitStmt_(const LetStmtNode* op) {
+  PrimExpr value = this->VisitExpr(op->value);
+  Stmt body = this->VisitStmt(op->body);
+  if (value.same_as(op->value) && body.same_as(op->body)) {
+    return GetRef<Stmt>(op);
+  } else if (value.dtype() == op->var->dtype) {
+    auto n = CopyOnWrite(op);
+    n->value = std::move(value);
+    n->body = std::move(body);
+    return Stmt(n);
+  } else {
+    auto new_var = op->var.copy_with_dtype(value.dtype());
+    Map<Var, PrimExpr> vmap{{op->var, new_var}};
+    auto new_body = SubstituteWithDataTypeLegalization(
+        std::move(body), [&](const Var& var) { return vmap.Get(var); });
+    // We need to visit the body again to insert additional casts
```

Comment:
it's possible to do this without revisiting. We can only visit value at the beginning, and update the vmap before visiting body

Reply:
We can pass `op->body` to `SubstituteWithDataTypeLegalization`, but we still need to visit `new_body` after `SubstituteWithDataTypeLegalization`. In that case, the output of the first visit,`this->VisitStmt(op->body)` at the top is only used for the first two `if / else` branches.

So this doesn't work, because the body of the Let has outdated variable dtype. Is this your understanding as well, or are you suggesting something else?
```
    auto new_body = SubstituteWithDataTypeLegalization(
        std::move(op->body), [&](const Var& var) { return vmap.Get(var); });
    // We need to visit the body again to insert additional casts
    return LetStmt(new_var, value, new_body, op->span);
```

Reply:
I'm experimenting with the following definition of `LetStmt` visitor, which looks better but it fails to pass the test for some reason (the cast before buffer store is not inserted). Looking.

```
Stmt DataTypeLegalizer::VisitStmt_(const LetStmtNode* op) {
  PrimExpr value = this->VisitExpr(op->value);
  auto body = op->body;
  auto new_var = op->var.copy_with_dtype(value.dtype());

  if (value.dtype() != op->var->dtype) {
    Map<Var, PrimExpr> vmap{{op->var, new_var}};
    body =
        SubstituteWithDataTypeLegalization(op->body, [&](const Var& var) { return vmap.Get(var); });
  }

  Stmt new_body = this->VisitStmt(body);

  if (value.same_as(op->value) && new_body.same_as(op->body)) {
    return GetRef<Stmt>(op);
  } else if (value.dtype() == op->var->dtype) {
    auto n = CopyOnWrite(op);
    n->value = std::move(value);
    n->body = std::move(body);
    return Stmt(n);
  } else {
    // We need to visit the body again to insert additional casts
    return LetStmt(new_var, value, new_body, op->span);
  }
}
```

Reply:
`SubstituteWithDataTypeLegalization` is implemented with `DataTypeLegalizer`. So I'm suggesting instead of calling `SubstituteWithDataTypeLegalization`, we just update the vmap before recursive visiting. Something like:
```
PrimExpr VisitExpr_(const VarNode* op) {
  if (auto it = vmap.find(op); it != vmap.end()) return (*it).second;
  return GetRef<Var>(op);
}

Stmt DataTypeLegalizer::VisitStmt_(const LetStmtNode* op) {
  PrimExpr new_value = this->VisitExpr(op->value);
  if (op->var->dtype != new_value.dtype()) {
    vmap[op->var] = new_var;
    new_body = VisitStmt(op->body);
    return LetStmt(new_var, new_value, new_body);
  }
}
```

Reply:
ok fixed, PTAL @vinx13

Reply:
sorry just saw your comment, looking now.

Reply:
@vinx13 Not sure where that `vmap` is coming from?

Reply:
oh I'm suggesting adding a new `vmap_` field

Reply:
I see, there is also `var_remap_` in the derived class, `IndexDataTypeNormalizer`. We may be able to unify

Reply:
done